File: go/2357.go
package main

import "container/heap"

type Heap []int

// Len implements heap.Interface
func (h *Heap) Len() int {
	return len(*h)
}

// Less implements heap.Interface
func (h *Heap) Less(i int, j int) bool {
	return (*h)[i] < (*h)[j]
}

// Swap implements heap.Interface
func (h *Heap) Swap(i int, j int) {
	(*h)[i], (*h)[j] = (*h)[j], (*h)[i]
}

// Pop implements heap.Interface
func (h *Heap) Pop() interface{} {
	ret := (*h)[len(*h)-1]
	*h = (*h)[:len(*h)-1]
	return ret
}

// Push implements heap.Interface
func (h *Heap) Push(x interface{}) {
	*h = append(*h, x.(int))
}

func (h *Heap) Top() int {
	return (*h)[0]
}

func minimumOperations(nums []int) int {
	var h *Heap = &Heap{}
	heap.Init(h)
	for _, v := range nums {
		if v != 0 {
			heap.Push(h, v)
		}
	}
	var ret int = 0
	for h.Len() > 0 {
		var x = heap.Pop(h)
		for h.Len() > 0 && h.Top() == x { // push until top != x
			heap.Pop(h)
		}
		ret++
	}
	return ret
}

File: cpp/2357.cpp
#include <functional>
#include <queue>
#include <vector>

using namespace std;

class Solution {
public:
  int minimumOperations(vector<int> &nums) {
    priority_queue<int, vector<int>, greater<int>> heap;
    for (int n : nums)
      if (n)
        heap.push(n);
    int ret = 0, x;
    while (heap.size()) {
      x = heap.top();
      heap.pop();
      while (heap.top() == x) {
        heap.pop();
        if (heap.size() == 0)
          break;
      }
      ret++;
    }
    return ret;
  }
};
